R/copy_dm_to.R

```diff
@@ -0,0 +1,169 @@
+## SQL script backend for copy_dm_to()
+## taken to new file to avoid merge conflicts to ongoing changes in db-interface.R #1739
+
+#' @name dm_sql
+#'
+#' @title Create \emph{DDL} and \emph{DML} scripts for `dm` and database connection
+#'
+#' @description
+#' Generate SQL scripts to create tables, load data and set constraints, keys and indices.
+#'
+#' @param dm A `dm` object.
+#' @param dest Connection to database.
+#' @param set_key_constraints If `TRUE` (default) will mirror `dm` primary and foreign key constraints on a database
+#'   and create unique indexes.
+#' @param table_names See argument description in \link{copy_dm_to}. Default to names of `dm`.
+#' @param temporary Should the tables be marked as \emph{temporary}? Defaults to `TRUE`.
+#' @param schema Name of schema to copy the `dm` to.
+#'
+#' @details
+#' \itemize{
+#'   \item{ `dm_ddl_pre` generates `CREATE TABLE` statements (including `PRIMARY KEY` definition). }
+#'   \item{ `dm_dml_load` generates `INSERT INTO` statements. }
+#'   \item{ `dm_ddl_post` generates scripts for `FOREIGN KEYS`, `UNIQUE KEYS` and `INDEXES`. }
+#'   \item{ `dm_sql` calls all three above and returns complete set of scripts. }
+#' }
+#'
+#' @return character vector of SQL statements.
+#'
+#' @export
+#' @examplesIf rlang::is_installed("RSQLite") && rlang::is_installed("dbplyr")
+#' con <- DBI::dbConnect(RSQLite::SQLite())
+#' dm <- dm_nycflights13()
+#' s <- dm_sql(dm, con)
+#' s
+#' DBI::dbDisconnect(con)
+dm_sql <- function(dm,
+                   dest,
+                   set_key_constraints = TRUE,
+                   table_names = set_names(names(dm)),
```

Comment:
Renaming tables (from `name` to `remote_name`) is more problematic because we lose original `name` inside `dm` object.
We still need it match autoincrement info from `dm_get_all_pks_impl()` because INSERT statements depends on autoincrement value.

Reply:
I pushed code towards this change, mostly commented out, in https://github.com/cynkra/dm/pull/1904/commits/636637f3e1081cddcb2e1d8a5c61788ddea78463

Another option is to introduce table "labels" or "remote_name" inside `dm` object. Even for local table.